server.go

```diff
@@ -140,8 +142,9 @@ func (server *mongoServer) AcquireSocket(poolLimit int, minPoolSize int, timeout
 			socket = &mongoSocket{
 				socketState: Connecting,
 			}
-			// hold our spot in the liveSockets slice
-			server.liveSockets = append(server.liveSockets, socket)
+			// hold a spot in the liveSockets slice to ensure connecting sockets are counted
+			// against the total connection cap.
+			server.liveSockets = append(server.liveSockets, nil)
```

Comment:
I would prefer the use of an atomicCounter for the livesocket count for bookkeeping, but that might need more code changes than this isolated one.

Reply:
I thought about that, but that requires additional coordination.  Since the following example is flacky when `n < limit - 1` and more than two processes are concurrently checking the limit:

```
	const limit = 8
	var p int64
	n := atomic.LoadInt64(&p)
	if n < limit && atomic.CompareAndSwapInt64(&p, n, n+1) {
		// do the thing
	} else {
		// if another process incremented p, we would need a 
		// bounded loop to do this properly and even that
		// isn't great since to be truly correct we would
		// also need a sync.Cond.
	}
```

Reply:
Basically, if we used atomics we'd end up with a sub-par mutex or a racy solution.

Reply:
Yeah, I don't think an atomic counter really helps us.

Normally I'd prefer we do separate explicit book keeping of sockets in connecting states rather than sticking nils into `liveSockets`, but given that the upstream project is dead and our limited expected usage and lack of good testing, I think the minimal surgical change is preferred.